### PR TITLE
Always re-run Feature migrations which have encountered errors

### LINKED ISSUE #83917: Feature migration refuses to run if all features requiring upgrade have encountered errors
### Elasticsearch Version

7.17.0

### Installed Plugins

none

### Java Version

_bundled_

### OS Version

Doesn't matter

### Problem Description

If all features which still require migration have encountered an error, trying to start migration produces no results, only a response claiming no migration is necessary, even if that's not true. This means that transient errors that occur during migration are impossible to retry.

The only known workaround for this is a full-cluster restart, which will clear the migration upgrade result metadata.

### Steps to Reproduce

1. Start a cluster using a 6.x version.
2. Upgrade it to 7.17.0.
3. Run system index migration, but ensure that some indices fail (e.g. by triggering https://github.com/elastic/elasticsearch/issues/83779, which `.tasks` always will - fortunately that is easy enough to work around by removing the index)
4. Try to re-run migration. Migration will not try to re-run.

### Logs (if relevant)

The response will just say no migration is necessary:
```json
{
  "accepted" : false,
  "reason" : "No system indices require migration"
}
```

### PR DESCRIPTION
**Pull request by AthenaEryma**
This PR addressed the behavior described in #83917, in which Feature migrations
which have encountered errors are not re-run in some cases. As of this PR, Features
which have encountered errors during migration are treated the same as Features 
requiring migration.

This PR also adds a test which artificially replicates #83917.

Fixes https://github.com/elastic/elasticsearch/issues/83917.

### CODE REVIEW

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java [hunk 1]
```diff
@@ -267,6 +270,62 @@ public void testMigrateIndexWithWriteBlock() throws Exception {
...
+                }
+
+                @Override
+                public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                    clusterStateUpdated.set(true);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    clusterStateUpdated.set(false);
```
Comment (rjernst) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java:305]: This should probably be captured and cause the assertBusy to fail with an assertion?

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java [hunk 2]
```diff
@@ -267,6 +270,62 @@ public void testMigrateIndexWithWriteBlock() throws Exception {
...
+                }
+            }, ClusterStateTaskExecutor.unbatched());
+
+        assertBusy(() -> assertTrue(Optional.ofNullable(clusterStateUpdated.get()).orElse(false)));
+
+        PostFeatureUpgradeRequest migrationRequest = new PostFeatureUpgradeRequest();
+        PostFeatureUpgradeResponse migrationResponse = client().execute(PostFeatureUpgradeAction.INSTANCE, migrationRequest).get();
+        // Make sure we actually started the migration
+        final Set<String> migratingFeatures = migrationResponse.getFeatures()
```
Comment (rjernst) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java:314]: Converting to a set isn't anymore efficient than just looking through the list. Is there som other reason we need an intermediate set to assert the feature was processed?

**Review by rjernst — COMMENTED** [review]
Looks good, just a couple comments on the test.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java (hunk 2 above)
Comment (gwbrown) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java:314]: This was a copy/paste from another test, you're right there's not much point to converting to a set here. I'll simplify.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java (hunk 1 above)
Comment (gwbrown) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/FeatureMigrationIT.java:305]: 👍 Good point.

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**